tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -0,0 +1,748 @@
+
+        /* Setup the client to have received a HelloRetryRequest */
+        memcpy_check(conn->secure.server_random, hello_retry_request_random_test_buffer, S2N_TLS_RANDOM_DATA_LEN);
+        conn->server_protocol_version = S2N_TLS13;
+        conn->client_protocol_version = S2N_TLS13;
+        conn->handshake.handshake_type = NEGOTIATED | FULL_HANDSHAKE | HELLO_RETRY_REQUEST;
+        conn->handshake.message_number = 1;
+        conn->secure.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+
+        EXPECT_SUCCESS(s2n_client_key_share_extension.send(conn, &key_share_extension));
+
+        const struct s2n_ecc_preferences *ecc_preferences = NULL;
+        EXPECT_SUCCESS(s2n_connection_get_ecc_preferences(conn, &ecc_preferences));
+        EXPECT_NOT_NULL(ecc_preferences);
+
+        /* should contain keyshare for only server negotiated curve */
+        for (size_t i = 0; i < ecc_preferences->count; i++) {
+            struct s2n_ecc_evp_params *ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
+            if (ecc_evp_params->negotiated_curve == conn->secure.server_ecc_evp_params.negotiated_curve) {
+                EXPECT_NOT_NULL(ecc_evp_params->negotiated_curve);
+                EXPECT_NOT_NULL(ecc_evp_params->evp_pkey);
+            } else {
+                EXPECT_NULL(ecc_evp_params->negotiated_curve);
+                EXPECT_NULL(ecc_evp_params->evp_pkey);
+            }
+        }
+
+        uint16_t key_shares_size;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&key_share_extension, &key_shares_size));
+        EXPECT_EQUAL(s2n_stuffer_data_available(&key_share_extension), key_shares_size);
+
+        /* should contain keyshare for only server negotiated curve */
+        uint32_t bytes_processed = 0;
+        EXPECT_EQUAL(key_shares_size, conn->secure.server_ecc_evp_params.negotiated_curve->share_size
+                                          + S2N_SIZE_OF_NAMED_GROUP + S2N_SIZE_OF_KEY_SHARE_SIZE);
+
+        uint16_t iana_value, share_size;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&key_share_extension, &iana_value));
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&key_share_extension, &share_size));
+        bytes_processed += conn->secure.server_ecc_evp_params.negotiated_curve->share_size + S2N_SIZE_OF_NAMED_GROUP
+                           + S2N_SIZE_OF_KEY_SHARE_SIZE;
+
+        EXPECT_EQUAL(iana_value, conn->secure.server_ecc_evp_params.negotiated_curve->iana_id);
+        EXPECT_EQUAL(share_size, conn->secure.server_ecc_evp_params.negotiated_curve->share_size);
+        EXPECT_SUCCESS(s2n_stuffer_skip_read(&key_share_extension, share_size));
+        EXPECT_EQUAL(bytes_processed, key_shares_size);
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&key_share_extension));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* For HelloRetryRequests, test that s2n_client_key_share_extension.recv can read and parse 
+     * the result of s2n_client_key_share_extension.send */
```

Comment:
Can you move all the client_key_share_extension tests to a file called client_key_share_extension_hrr_tests or something? I'm not sure they belong with the client hello tests.